ambry-store/src/test/java/com.github.ambry.store/BlobStoreStatsTest.java

```diff
@@ -0,0 +1,335 @@
+    long totalContainerValidSizeAfterDeletes = verifyContainerValidSize();
+    long expectedLogSegmentDecrement =
+        2 * (CuratedLogIndexState.PUT_RECORD_SIZE - CuratedLogIndexState.DELETE_RECORD_SIZE);
+    long expectedContainerDecrement = 2 * CuratedLogIndexState.PUT_RECORD_SIZE;
+    assertEquals("Delete entries are not properly counted for log segment valid size",
+        totalLogSegmentValidSizeAfterDeletes, totalLogSegmentValidSizeBeforeDeletes - expectedLogSegmentDecrement);
+    assertEquals("Delete entries are not properly counted for container valid size",
+        totalContainerValidSizeAfterDeletes, totalContainerValidSizeBeforeDeletes - expectedContainerDecrement);
+  }
+
+  /**
+   * Verify the correctness of valid data size information per container returned by BlobStoreStats.
+   * @return the total valid data size of all containers (from all serviceIds)
+   */
+  private long verifyContainerValidSize() throws StoreException {
+    long referenceTimeInMs = state.time.milliseconds();
+    Map<String, Map<String, Long>> actualContainerValidSizeMap = blobStoreStats.getValidDataSizeByContainer();
+    Map<String, Map<String, Long>> expectedContainerValidSizeMap = getValidSizeByContainer(referenceTimeInMs);
+    long totalValidSize = 0L;
+
+    for (Map.Entry<String, Map<String, Long>> expectedContainerValidSizeEntry : expectedContainerValidSizeMap.entrySet()) {
+      String serviceId = expectedContainerValidSizeEntry.getKey();
+      assertTrue("Expected serviceId: " + serviceId + " not found", actualContainerValidSizeMap.containsKey(serviceId));
+      Map<String, Long> innerMap = expectedContainerValidSizeEntry.getValue();
+      for (Map.Entry<String, Long> innerEntry : innerMap.entrySet()) {
+        String containerId = innerEntry.getKey();
+        assertTrue("Expected containerId: " + containerId + " not found in serviceId: " + serviceId,
+            innerMap.containsKey(containerId));
+        long expectedContainerValidSize = innerEntry.getValue();
+        long actualContainerValidSize = actualContainerValidSizeMap.get(serviceId).get(containerId);
+        assertEquals("Valid data size mismatch for serviceId: " + serviceId + " containerId: " + containerId,
+            expectedContainerValidSize, actualContainerValidSize);
+        totalValidSize += expectedContainerValidSize;
+      }
+    }
+
+    return totalValidSize;
+  }
+
+  /**
+   * Verify the correctness of valid data size information per log segment returned by BlobStoreStats for a given {@link TimeRange}.
+   * @param timeRange the {@link TimeRange} to be used for the verification
+   * @return the total valid data size of all log segments
+   */
+  private long verifyLogSegmentValidSize(TimeRange timeRange) throws StoreException {
+    Pair<Long, NavigableMap<String, Long>> actualLogSegmentValidSizeMap = blobStoreStats.getValidDataSizeByLogSegment(timeRange);
```

Comment:
take exceptionRefTime here itslef. Don't use diff timeRefs for different log segments as in line 263